Add G_k(r) type to PDFFourierTransform2

sf1919 opened the pull request:
**Description of work.**
Added a new PDF type of G_k(r) to the PDFFourierTransform (v2) algorithm. Further discussions identified that the formula provided in the issue was incorrect and it has now been revised to the following:

$$
G_k(r) = \frac{0.01 * \langle b_{coh} \rangle ^2 G^{PDF}(r)}{(4 \pi)^2 \rho_0 r} = 0.01 * \langle b_{coh} \rangle ^2 [g(r)-1]
$$

This PR adds the functionality required to use the new PDF Type, documentation and relevant tests.

**To test:**

Unzip this zip file somewhere: 
[POLARISPowderDiffractionFolderSmall.zip](https://github.com/mantidproject/mantid/files/9103459/POLARISPowderDiffractionFolderSmall.zip)

Update the paths in the file polaris_config_example.yaml so they match where you unzipped the file and also update config_file_path in the script below

Run the following script in workbench and check it completes. In addition plot the PDF workspace and check that it is non-zero:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder import polaris, SampleDetails

config_file_path = r"C:\Total Scattering\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)

polaris.focus(run_number="98533", input_mode='Summed')

q_lim = [[2.5, 3, 4, 6, 7], [3.5, 5, 7, 11, 40]]

polaris.create_total_scattering_pdf(run_number="98533", merge_banks=True,q_lims=q_lim, 
                                    freq_params=[1.4], pdf_type="G_k(r)", debug=True)
```

Fixes #34014 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Added fwd G_k(r) functionality to PDFFourierTransform2

sf1919 committed:
Add bkwd G_k(r) functionality to PDFFourierTransform2

sf1919 committed:
Added documentation

sf1919 committed:
Add error calculations

sf1919 committed:
Revised formula used

- Need to multiply by a factor of 0.001 * cohScatterLength of the material instead of 4pi
- Updated documentation with revised formulae

sf1919 committed:
Added 2 tests

- added a test to check that calling the different PDF Types will execute the algorithm
- added a test to check the values are expected for each of the different PDF Types in Forward calculations

sf1919 committed:
Moved getProperty calls out of PDFType methods

- Changed convertToLittleGRMinus1 and convertFromLittleGRMinus1 so values are passed to the method rather than utilising calls to the Properties within the method.

sf1919 committed:
Added test for PDFTypes when direction=Backward

DannyHindson commented on Framework/Algorithms/test/PDFFourierTransform2Test.h:
```diff
@@ -260,6 +289,132 @@ class PDFFourierTransform2Test : public CxxTest::TestSuite {
+    const double single_x = 2.0;
+
+    // set up initial y values
+    std::vector<double> y(2, 5.0);
```
you could use single_x here I think instead of repeating the value 2

DannyHindson replied:
Just spotted your latest commit. My comment here was wrong because the 2 here is the size. Was thinking it was somehow an (x,y) value pair which it isn't

DannyHindson reviewed (changes requested):
Hopefully you'll pick this up on Monday - certainly not one for this afternoon\evening!
I've added a few detail level things but nice job overall. All pretty trivial to fix I think.
Also did you see my Slack message from a few days ago about the list of pdf types that exists in the unit polaris.py:
```
        if 'pdf_type' not in kwargs or kwargs['pdf_type'] not in ['G(r)', 'g(r)', 'RDF(r)']:
            kwargs['pdf_type'] = 'G(r)'
            logger.warning('PDF type not specified or is invalid, defaulting to G(r)')
```
You need to add the new one into that list as well (and that's the reason the PDF plot for small values of r wasn't what I was expecting when we did a screen share a few days back)

sf1919 committed:
Revisions following review

- updated polaris script
- made better use of variables
- correced maths error (replaced 0.001 with 0.01 where necessary)
- changed use of double_t to double
- renamed method in test that runs the algorithm

sf1919 committed:
Fixed issue with test

DannyHindson commented on Framework/Algorithms/test/PDFFourierTransform2Test.h:
```diff
@@ -260,6 +289,132 @@ class PDFFourierTransform2Test : public CxxTest::TestSuite {
+    // test G(r) - note that y values have been changed by previous call to convertToLittleGRMinus1
+    const std::string BIG_G_OF_R("G(r)");
+    pdfft.convertToLittleGRMinus1(y, x, dy, dx, BIG_G_OF_R, rho0, cohScatLen);
+    const auto exp_big_gr = exp_gr / (factor1 * single_x);
```
Couple of things relating to these backward tests:

1) please can you rename this `exp_big_gr` variable. It actually holds a quantity little g(r). So strictly speaking it is `exp_little_gr_that_has_been_obtained_from_big_g_r_input` but since that's a bit of a mouthful perhaps simply calling it `exp_little_gr` is fine?

2) Please can you have a go at not chaining together the tests? The fact that the y values output from the previous test get used as the input for the next one is a bit confusing. 

If you declare your initial vector higher up on L357 as:
```
std::vector<double> y_initial(2, 5.0);
```
...and then take a copy of it before each test it should work. So for the following test do something like:
```
auto y = y_initial;
pdfft.convertToLittleGRMinus1(y, x, dy, dx, ...);
const auto exp_big_gr = y_initial / (factor1 * single_x);
```
so I don't think this will inolve changing all the formulae that you have for your exp_xxx variables

DannyHindson reviewed (changes requested):
Latest changes look good. Just noticed there's a place in PDFFourierTransform-v2.rst where the new PDF type also needs a mention (see ~L15):
```
The available PDF types are the
reduced pair distribution function :math:`G(r)`, the pair distribution function :math:`g(r)`, and the
radial distribution function :math:`RDF(r)`.
```
If my comments about the backward test don't make sense then fine to get together on Zoom to go through the changes so we can round this off

sf1919 committed:
Further changes after review

- Updated test_PDFTypes_bkwd and documentation

DannyHindson commented on Framework/Algorithms/src/PDFFourierTransform2.cpp:
```diff
@@ -399,11 +419,15 @@ void PDFFourierTransform2::exec() {
 
+  double rho0 = determineRho0();
+  const Kernel::Material &material = inputWS->sample().getMaterial();
+  const auto cohScatLen = material.cohScatterLength();
```
I think that if the cohScatLen is zero and PDFType=G_k here we should throw an exception with a message saying "Coherent Scattering Length is zero. Please check a sample material has been specified" or similar.
This will involve you storing getProperty("PDFType") in a variable as well I think
You would get this error condition if you ran this script which is a variation on the usage example in the algorithm documentation. At the moment the algorithm just runs with cohScatLen=0 and produces a zero output:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# Simulates Load of a workspace with all necessary parameters
import numpy as np;
xx = np.array(range(0,100))*0.1
yy = np.exp(-(2.0 * xx)**2)
yy = np.delete(yy,-1) # need one less Y value than X value for histogram data
ws = CreateWorkspace(DataX=xx, DataY=yy, UnitX='MomentumTransfer')
Rt = PDFFourierTransform(ws, SofQType='S(Q)-1', PDFType='G_k(r)')

# Look at sample results:
print('part of S(Q)-1 and its correlation function')
for i in range(10):
   print('! {0:4.2f} ! {1:5f} ! {2:f} ! {3:5f} !'.format(xx[i], yy[i], Rt.readX(0)[i], Rt.readY(0)[i]))
```

DannyHindson reviewed (changes requested):
The changes to unchain the backward tests look good. I've just added some comments on the variable names - but functionally they are correct.
I've also spotted an error condition that I think we should catch where a user hasn't specified a material
Shall we get together on a Zoom call later to go through these last couple of things?

sf1919 committed:
Additional changes on review

- Removal of obsolete comments
- Renaming variables in test_PDFTypes_bkwd
- Added new validation for when material not provided when calculating G_k(r)
- Updated test_check_PDFType_execute to work with new validation

DannyHindson reviewed (approved):
Thanks for the changes. All looks good now
